.travis.yml

```diff
@@ -14,24 +14,147 @@ git:
+      env: GCC=-4.8 MPI=none
+    -
+      <<: *foureight
+      env: GCC=-4.8 MPI=MPICH
+    - &fournine
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-4.9', 'g++-4.9', 'gfortran-4.9', 'ccache']
+      env: GCC=-4.9 MPI=none
+    -
+      <<: *fournine
+      env: GCC=-4.9 MPI=MPICH
+    - &five
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-5', 'g++-5', 'gfortran-5', 'ccache']
+      env: GCC=-5 MPI=none
+
+    -
+      <<: *five
+      env: GCC=-5 MPI=MPICH
+    - &six
+      <<: *baseconf
+      addons:
+        apt:
+          sources: ['ubuntu-toolchain-r-test']
+          packages: ['gcc-6', 'g++-6', 'gfortran-6', 'ccache']
+      env: GCC=-6 MPI=none
+    -
+      <<: *six
+      env: GCC=-6 MPI=MPICH
 
 before_install:
-  - openssl aes-256-cbc -K $encrypted_c48b4c75cf28_key -iv $encrypted_c48b4c75cf28_iv -in id_rsa_Travis2GH.enc -out id_rsa_Travis2GH -d # encrypted with `travis encrypt-file`
-  - unset -v ${!encrypted_@} # unset secret encrypted variables
-  - chmod 600 id_rsa_Travis2GH
+  - if [[ ! -d "${HOME}/.ccache" ]] ; then mkdir "${HOME}/.ccache" ; else ls -a "${HOME}/.ccache" ; fi
+  - ccache --zero-stats
   - if [[ ! -d "${HOME}/bin" ]] ; then mkdir "${HOME}/bin" && export PATH="${HOME}/bin:${PATH}"; fi
-  - printf '#!/usr/bin/env bash\nssh -i %s/id_rsa_Travis2GH "$@"\n' "$(pwd)" > "${HOME}/bin/ssh_to_GH"
-  - chmod +x "${HOME}/bin/ssh_to_GH"
-  - export GIT_SSH="${HOME}/bin/ssh_to_GH"
-  - echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+  - if [[ ! -d "${HOME}/mpich/build" ]] ; then mkdir -p "${HOME}/mpich/build" ; fi
+  - if [[ ! -d "${HOME}/usr" ]] ; then mkdir -p "${HOME}/usr/{bin,lib,share,include,lib64,libexec}" && export PATH="${HOME}/usr/bin:${PATH}"; fi
+  - |
+    if [[ -n "${encrypted_c48b4c75cf28_key}" ]] ; then # not a PR or someone else's repo
+      openssl aes-256-cbc -K $encrypted_c48b4c75cf28_key -iv $encrypted_c48b4c75cf28_iv -in id_rsa_Travis2GH.enc -out id_rsa_Travis2GH -d \
+      && chmod 600 id_rsa_Travis2GH \
+      && printf '#!/usr/bin/env bash\nssh -i %s/id_rsa_Travis2GH "$@"\n' "$(pwd)" > "${HOME}/bin/ssh_to_GH" \
+      && chmod +x "${HOME}/bin/ssh_to_GH" \
+      && export GIT_SSH="${HOME}/bin/ssh_to_GH" \
+      && echo "Testing ssh to github.com" && ($GIT_SSH -T "git@github.com" || true)
+    fi
+  - unset -v ${!encrypted_@} # unset secret encrypted variables
   - git --version
   - env | grep -iv 'token\|key\|encrypt' | sort # scrub tokens from output & sort
   - mount
+  - ls -l $(which gcc) && gcc --version # dump info about compilers
+  - ls -l $(which g++) && g++ --version
+  - if which gfortran; then ls -l $(which gfortran) && gfortran --version ; fi
+  - if ((RANDOM % 2 == 0)); then SUF="${GCC:-}" ; else SUF="" ; fi # randomize keeping suffix on links
```

Comment:
Randomize adding suffixes to links in `~/bin` since sometimes tau doesn't like this: #109